### ISSUE TITLE
Include directory should override exclude file

### ISSUE THREAD

**Issue by axesider** [cpburnz/python-pathspec#74]
```
import pathspec

spec = pathspec.GitIgnoreSpec.from_lines([
    '*',  # Ignore all files by default
    '!*/',  # but scan all directories
    '!*.txt',  # Text files
    '/test1/**'  # ignore all in the directory
])
files = {
    'test1/b.bin',
    'test1/a.txt',
    'test1/c/c.txt',
    'test2/a.txt',
    'test2/b.bin',
    'test2/c/c.txt',
}
ignores = set(spec.match_files(files))
print(ignores)
```
>{'test1/b.bin', 'test2/b.bin'}

It should be
>{'test1/a.txt', 'test1/c/c.txt', 'test1/b.bin', 'test2/b.bin'}

I think [GitIgnoreSpec._match_file](https://github.com/cpburnz/python-pathspec/blob/8a7c8fc5a51c81e4d226251405f3aebd669b8450/pathspec/gitignore.py#L131) should have something like:

```
                if pattern.include is False and dir_mark and out_priority == 0:
                    out_matched = pattern.include
                    out_priority = priority
                elif pattern.include is True and dir_mark:
                    out_matched = pattern.include
                    out_priority = priority
                elif priority >= out_priority:
                    out_matched = pattern.include
                    out_priority = priority
```

**Comment by cpburnz** [comment]
I've confirmed Git behaves the way you described and your suggested fix does fix the discrepancy. Thanks.

What was the intended purpose of the first if-statement? `if pattern.include is False and dir_mark and out_priority == 0:` can only evaluate to true for the first matching pattern, and it must be excluding a directory pattern. The first matching pattern will always have its `out_matched` and `out_priority` set (from `elif priority >= out_priority:`) so think that makes the first if-statement redundant.

**Comment by bkarstens** [comment]
@axesider is `!*/` a special case? if I anchor all the paths into folder `foo`, the results change and I don't know why; see lines 9 and 12:
![image](https://user-images.githubusercontent.com/17677688/224535088-90b79bc8-24c4-4e17-9e8b-eeb765657a51.png)

**Comment by cpburnz** [comment]
This fix is now in release v0.11.1.
